File: mergeSort.c
#include <iostream>
using namespace std;

void merge(int *,int, int , int );
void mergesort(int *a, int low, int high)
{
    int mid;
    if (low < high)
    {
mid=(low+high)/2;
mergesort(a,low,mid);
mergesort(a,mid+1,high);
merge(a,low,high,mid);
    }
    return;
}
// Merge sort concepts starts here
void merge(int *a, int low, int high, int mid)
{
    int i, j, k, c[50];
    i = low;
    k = low;
    j = mid + 1;
    while (i <= mid && j <= high)
    {
if (a[i] < a[j])
{
    c[k] = a[i];
    k++;
    i++;
}
else
{
    c[k] = a[j];
    k++;
    j++;
}
    }
    while (i <= mid)
    {
c[k] = a[i];
k++;
i++;
    }
    while (j <= high)
    {
c[k] = a[j];
k++;
j++;
    }
    for (i = low; i < k; i++)
    {
a[i] = c[i];
    }
}
// from main mergesort function gets called
int main()
{
    int a[30], i, b[30];
    cout<<"enter  the number of elements:";
    for (i = 0; i <= 5; i++) { cin>>a[i];
    }
    mergesort(a, 0, 4);
    cout<<"sorted array\n";
    for (i = 0; i < 5; i++)
    {
cout<<a[i]<<"\t";
    }
    cout<<"enter  the number of elements:";
    for (i = 0; i < 5; i++) { cin>>b[i];
    }
    mergesort(b, 0, 4);
    cout<<"sorted array:\n";
    for (i = 0; i < 5; i++)
    {
cout<<b[i]<<"\t";
    }
    getch();
}

File: MergeSort.java
package mergesort;

public class MergeSort {
 
void mge(int ar[], int st, int mi, int en)  
{  
  
int l = mi - st + 1;  
int r = en - mi;  
  
int LArray[] = new int [l];  
int RArray[] = new int [r];  
  
for (int i=0; i<l; ++i)  
LArray[i] = ar[st + i];  
  
for (int j=0; j<r; ++j)  
RArray[j] = ar[mi + 1+ j];  
  
  
int i = 0, j = 0;  
int k = st;  
while (i<l&&j<r)  
{  
if (LArray[i] <= RArray[j])  
{  
ar[k] = LArray[i];  
i++;  
}  
else  
{  
ar[k] = RArray[j];  
j++;  
}  
k++;  
}  
while (i<l)  
{  
ar[k] = LArray[i];  
i++;  
k++;  
}  
  
while (j<r)  
{  
ar[k] = RArray[j];  
j++;  
k++;  
}  
}  
  
void sort(int ar[], int st, int en)  
{  
if (st<en)  
{  
int mi = (st+en)/2;  
sort(ar, st, mi);  
sort(ar , mi+1, en);  
mge(ar,st,mi,en);  
}  
}  
public static void main(String args[])  
{  
int ar[] = {15,60,67,3,54,45,1,89,77,22,16,44,54};  
MergeSort ob = new MergeSort();  
ob.sort(ar, 0, ar.length-1);  
  
System.out.println("\nSorted array");  
for(int i =0; i<ar.length;i++)  
{  
    System.out.println(ar[i]+"");  
}  
}  
}  

